integrations/access/accesslist/app.go

```diff
@@ -0,0 +1,251 @@
+	notificationStart := accessList.Spec.Audit.NextAuditDate.Add(-accessList.Spec.Audit.Notifications.Start)
+
+	// If the current time before the notification start time, skip notifications.
+	if now.Before(notificationStart) {
+		log.Infof("Access list %s is not ready for notifications, notifications start at %s", accessList.GetName(), notificationStart.Format(time.RFC3339))
+		return nil
+	}
+
+	// Get the owners from the bot as recipients.
+	for _, owner := range accessList.Spec.Owners {
+		recipient, err := a.bot.FetchRecipient(ctx, owner.Name)
+		if err != nil {
+			log.Debugf("error getting recipient %s", owner.Name)
+			continue
+		}
+		allRecipients[owner.Name] = *recipient
+	}
+
+	if len(allRecipients) == 0 {
+		return trace.NotFound("no recipients could be fetched for access list %s", accessList.GetName())
+	}
+
+	// Try to create base notification data with a zero notification date. If these objects already
+	// exist, that's okay.
+	userNotifications := map[string]time.Time{}
+	for _, recipient := range allRecipients {
+		userNotifications[recipient.Name] = time.Time{}
+	}
+	_, err := a.pluginData.Create(ctx, accessList.GetName(), pd.AccessListNotificationData{
+		UserNotifications: userNotifications,
+	})
+
+	// Error is okay so long as it's already exists.
+	if err != nil && !trace.IsAlreadyExists(err) {
+		return trace.Wrap(err, "during create")
+	}
+
+	// Calculate weeks from start.
+	weeksFromStart := now.Sub(notificationStart) / oneWeek
+	windowStart := notificationStart.Add(weeksFromStart * oneWeek)
+
+	recipients := []recipient.Recipient{}
+	_, err = a.pluginData.Update(ctx, accessList.GetName(), func(data pd.AccessListNotificationData) (pd.AccessListNotificationData, error) {
+		userNotifications := map[string]time.Time{}
+		for _, recipient := range allRecipients {
+			lastNotification := data.UserNotifications[recipient.Name]
+
+			// If the notification window is before the last notification date, then this user doesn't need a notification.
+			if !windowStart.After(lastNotification) {
+				log.Infof("User %s has already been notified", recipient.Name)
+				userNotifications[recipient.Name] = lastNotification
+				continue
+			}
+
+			recipients = append(recipients, recipient)
+			userNotifications[recipient.Name] = now
+		}
+		return pd.AccessListNotificationData{UserNotifications: userNotifications}, nil
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if len(recipients) == 0 {
+		log.Infof("Nobody to notify for access list %s", accessList.GetName())
+		return nil
+	}
+
+	return trace.Wrap(a.bot.SendReviewReminders(ctx, recipients, accessList))
+}
+
+// accessListClient will return an access list client for the plugin manager.
+func accessListClient(client teleport.Client) services.AccessListsGetter {
+	type accessListClient[T services.AccessListsGetter] interface {
+		AccessListClient() T
+	}
+
+	switch client := client.(type) {
+	case accessListClient[services.AccessLists]:
```

Comment:
This is pretty hard for me to reason about, and in general it's not the kind of code anyone likes to maintain (it doesn't _do_ anything, it just appeases the compiler).

I'll ask the same question here - do we really need the generics? Is there a way we can simplify this?

Reply:
The api client can be used and `auth.ClientI` can be used for the plugin's custom Teleport client interface. Though both of these implement `AccessListClient()`, the API client returns `*accesslist.Client`. This implements `services.AccessLists`, but since it doesn't return `services.AccessLists`, it doesn't technically conform to `AccessListClient() services.AccessLists`. I can get rid of the generic here through some wizardry, it looks like. I'll do my best to excise the generics here overall.